File: cpp/401_500/409_longest_palindrome.cpp
//
//  409_longest_palindrome.cpp
//  leetcode
//
//  Created by Hang Zhang on 6/29/17.
//  Copyright © 2017 Hilings Studio. All rights reserved.
//

#include <iostream>
using namespace std;

class Solution {
public:
    int longestPalindrome(string s) {
        int count[52] = {0}, odd = 0, len = 0;
        for (char c: s) {
            if (c >= 'A' && c <= 'Z') {
                count[c-'A']++;
            } else {
                count[26+c-'a']++;
            }
        }
        for (int a: count) {
            if (a % 2 == 0) {
                len += a;
            } else {
                len += a - 1;
                odd = 1;
            }
        }
        return len+odd;
    }
};

int main(int arg, char *argv[]) {
    // insert code here...
    cout << "LeetCode 409. Longest Palindrome, C++ ...\n\n";
    Solution sol;
    
    string s = "abccccdd";	// 7
    s = "ccc";	// 3
    int r = sol.longestPalindrome(s);
    cout << r << '\n';
    
    return 0;
}

File: go/401-500/409_Longest_Palindrome.go
package main

func longestPalindrome(s string) int {
	m := map[rune]int{} // count per char
	for _, c := range s {
		m[c]++
	}

	r := 0
	single := 0 // any char with odd count
	for _, count := range m {
		r += count / 2 * 2
		if count%2 == 1 {
			single = 1
		}
	}
	return r + single
}

// func main() {
// 	fmt.Printf("LeetCode 409. Longest Palindrome ...\n\n")

// 	type args struct {
// 		s string
// 	}
// 	tests := []struct {
// 		args args
// 		want int
// 	}{
// 		{
// 			args: args{
// 				s: "abccccdd",
// 			},
// 			want: 7,
// 		},
// 		{
// 			args: args{
// 				s: "a",
// 			},
// 			want: 1,
// 		},
// 		{
// 			args: args{
// 				s: "bb",
// 			},
// 			want: 2,
// 		},
// 		{
// 			args: args{
// 				s: "ccc",
// 			},
// 			want: 3,
// 		},
// 	}
// 	for _, tt := range tests {
// 		r := longestPalindrome(tt.args.s)
// 		fmt.Printf("got = %v, want = %v\n\n", r, tt.want)
// 	}
// }
